### PR TITLE
Bed 4069

### PR DESCRIPTION
**Pull request by benwaples**
## Description
Adds `Partially Completed` status for analysis to reflect when parts of analysis fail and others do not. This means that analysis can complete 1 of 3 ways:
1. `Complete` - All of analysis completes successfully
2. `Partially Complete` - Some parts of analysis failed, however at least one was successful.
3. `Failed` - All parts of analysis failed.

## Motivation and Context
When customers go to see the status of collection and analysis, they are only advised `Complete` or `Failed`.  This is based on the complete analysis finishing without any warnings or errors. Because some of analysis can fail but some data is still processed, a status of Partially Completed with additional logging should help trouble shoot the problem.

## How Has This Been Tested?
Manual testing errors happening at each step of analysis independently and collectively.

## Types of changes
-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
-   [ ] Documentation updates are needed, and have been made accordingly.
-   [ ] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My changes include a database migration.

### CODE REVIEW

File: cmd/api/src/daemons/datapipe/analysis.go [hunk 1]
```diff
@@ -33,57 +34,85 @@ import (
 	"github.com/specterops/bloodhound/src/services/dataquality"
 )
 
+var (
+	ErrAnalysisFailed             = errors.New("analysis failed")
+	ErrAnalysisPartiallyCompleted = errors.New("analysis partially failed")
```
[the remark this answers is not in the document]
Comment (benwaples) [reply cmd/api/src/daemons/datapipe/analysis.go:39]: thats a great point! Ill fix that

File: cmd/api/src/daemons/datapipe/analysis.go [hunk 2]
```diff
@@ -33,57 +34,81 @@ import (
...
+
 func RunAnalysisOperations(ctx context.Context, db database.Database, graphDB graph.Database, _ config.Configuration) error {
-	var (
-		collector = &errors.ErrorCollector{}
-	)
+	var collectedErrors error
 
 	if err := adAnalysis.FixWellKnownNodeTypes(ctx, graphDB); err != nil {
-		collector.Collect(fmt.Errorf("fix well known node types failed: %w", err))
+		collectedErrors = errors.Join(collectedErrors, fmt.Errorf("fix well known node types failed: %w", err))
```
Comment (computator) [inline cmd/api/src/daemons/datapipe/analysis.go:46]: This is not the intended usage of `errors.Join()`. It is not like `append()` and repeatedly calling it to add a new error will not end up with a list, but rather somewhat of a nested tree structure. `errors.Join()` is intended to be called once with the final list of errors. (See: https://github.com/golang/go/issues/60209 and implementation in https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/errors/join.go;l=15)

This also applies to many of the other uses throughout the PR

**Review by computator — CHANGES_REQUESTED** [review]
This looks good overall, but I added some feedback on the implementation

File: cmd/api/src/daemons/datapipe/jobs.go [hunk 3]
```diff
@@ -54,6 +54,24 @@ func FailAnalyzedFileUploadJobs(ctx context.Context, db database.Database) {
...
+	if ctx.Err() != nil {
+		return
+	}
+
+	if fileUploadJobsUnderAnalysis, err := db.GetFileUploadJobsWithStatus(model.JobStatusAnalyzing); err != nil {
+		log.Errorf("Failed to load file upload jobs under analysis: %v", err)
+	} else {
+		for _, job := range fileUploadJobsUnderAnalysis {
+			if err := fileupload.UpdateFileUploadJobStatus(db, job, model.JobStatusPartiallyComplete, "Analysis Partially Completed"); err != nil {
+				log.Errorf("Failed updating file upload job %d to failed status: %v", job.ID, err)
```
[the remark this answers is not in the document]
Comment (benwaples) [reply cmd/api/src/daemons/datapipe/jobs.go:69]: ahh good catch! I can update that. Thanks

File: cmd/api/src/daemons/datapipe/analysis.go (hunk 2 above)
Comment (computator) [reply cmd/api/src/daemons/datapipe/analysis.go:46]: We discussed this in Slack, but my suggestion is to `append()` to a slice of errors and then you can use `errors.Join()` to combine them all one time at the end.

**Review by computator — APPROVED** [review]

**Pull request merged**